spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -173,11 +172,13 @@ public void replaceChannelMappings(Properties channelMappings) {
 	}
 
 	private void doSetChannelMappings(Map<String, String> newChannelMappings) {
-		Map<String, String> oldChannelMappings = this.channelMappings;
-		this.channelMappings = newChannelMappings;
+		Map<String, String> oldChannelMappings = new HashMap<String, String>(this.channelMappings);
+		synchronized (this.channelMappings) {
```

Comment:
We have to trust that all (and future) subclasses synchronize on the map but I don't think we want the overhead of copying it for every message.

This is why I prefer the atomic update approach.

I think it would be much better to change `setChannelMapings` and `removeChannelMapping` to use an atomic update as well, instead of updating in-place.